### PR TITLE
chakrashim: Implementing v8::Message

### PR DESCRIPTION
**Pull request by MSLaguana**
Using the new JSRT method JsGetAndClearExceptionWithMetadata we can implement
more v8::Message functionality which lets us give better messages for
unhandled exceptions.

##### Checklist

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)

chakrashim

### CODE REVIEW

File: deps/chakrashim/src/v8trycatch.cc [hunk 1]
```diff
@@ -78,23 +78,34 @@ void TryCatch::GetAndClearException() {
...
+  JsPropertyIdRef err = jsrt::IsolateShim::GetCurrent()
+    ->GetCachedPropertyIdRef(jsrt::CachedPropertyIdRef::exception);
+  if (err == JS_INVALID_REFERENCE) {
+    return Local<Value>();
+  }
+
+  JsValueRef error;
+  if (JsGetProperty(metadata, err, &error) != JsNoError) {
     return Local<Value>();
   }
```
Comment (agarwal-sandeep) [inline deps/chakrashim/src/v8trycatch.cc]: Same code is in 3 functions ReThrow, Exception, StackTrace. Can we refactor and have it in one private function and call that in all these functions?

File: deps/chakrashim/src/v8message.cc [hunk 2]
```diff
@@ -28,8 +28,18 @@ Local<String> Message::Get() const {
 }
 
 MaybeLocal<String> Message::GetSourceLine(Local<Context> context) const {
-  // CHAKRA-TODO: Figure out how to transmit this info...?
-  return Local<String>();
+  JsPropertyIdRef source = jsrt::IsolateShim::GetCurrent()
+    ->GetCachedPropertyIdRef(jsrt::CachedPropertyIdRef::source);
```
Comment (kfarnung) [inline deps/chakrashim/src/v8message.cc]: nit:  Elsewhere in the node code they use double-indent for continuations from the previous line to help it stand out.

File: deps/chakrashim/src/v8message.cc [hunk 3]
```diff
@@ -28,8 +28,18 @@ Local<String> Message::Get() const {
...
+  if (source == JS_INVALID_REFERENCE) {
```
Comment (kfarnung) [inline deps/chakrashim/src/v8message.cc]: In what cases can this happen?  If it's only coding errors it seems better to just fast fail using CHAKRA_VERIFY or similar.  That said, I see a lot of examples where the result is just trusted and passed on, so maybe the check is not necessary?

File: deps/chakrashim/src/v8message.cc [hunk 4]
```diff
@@ -41,37 +51,87 @@ ScriptOrigin Message::GetScriptOrigin() const {
...
-  // CHAKRA-TODO: Figure out how to transmit this info...?
-  return Nothing<int>();
+  JsPropertyIdRef columnProp = jsrt::IsolateShim::GetCurrent()
+    ->GetCachedPropertyIdRef(jsrt::CachedPropertyIdRef::column);
+  if (columnProp == JS_INVALID_REFERENCE) {
+    return Nothing<int>();
+  }
+
+  JsValueRef result;
+  if (JsGetProperty((JsValueRef)this, columnProp, &result) != JsNoError) {
```
Comment (kfarnung) [inline deps/chakrashim/src/v8message.cc]: You should be able to use jsrt::ValueToIntLikely

**Review by kfarnung — APPROVED** [review]
🚢 LGTM with minor fixes

File: test/parallel/test-vm-syntax-error-stderr.js [hunk 5]
```diff
@@ -23,13 +23,8 @@ p.stderr.on('data', function(data) {
...
-    assert(/^SyntaxError: Expected ';'/.test(output));
-  } else {
-    assert(/BEGIN CERT/.test(output));
-    assert(/^\s+\^/m.test(output));
-    assert(/Invalid left-hand side expression in prefix operation/
-      .test(output));
-  }
+  assert(/BEGIN CERT/.test(output));
+  assert(/^\s+\^/m.test(output));
+  assert(/Error:/
```
Comment (kunalspathak) [inline test/parallel/test-vm-syntax-error-stderr.js]: >/Error:/ [](start = 9, length = 8)

can you make sure that this test passes with v8?

Comment (MSLaguana) [reply test/parallel/test-vm-syntax-error-stderr.js]: It does.

Comment (kunalspathak) [reply test/parallel/test-vm-syntax-error-stderr.js]: actually, just checked upstream and your change diverges the upstream test case. Can you wrap this error under `common.engineSpecificMessage()`?

---
In reply to: [118543511](https://github.com/nodejs/node-chakracore/pull/263#discussion_r118543511) [](ancestors = 118543511)

File: test/message/vm_display_runtime_error.chakracore.out [hunk 6]
```diff
@@ -1,4 +1,8 @@
 beginning
+vm.js:44
```
Comment (kunalspathak) [inline test/message/vm_display_runtime_error.chakracore.out]: >44 [](start = 6, length = 2)

out file takes regex. So can you please replace all the line numbers with `*`?

File: deps/chakrashim/src/v8trycatch.cc [hunk 7]
```diff
@@ -78,23 +78,34 @@ void TryCatch::GetAndClearException() {
...
 Handle<Value> TryCatch::ReThrow() {
-  if (error == JS_INVALID_REFERENCE) {
+  if (metadata == JS_INVALID_REFERENCE) {
     GetAndClearException();
   }
 
-  if (error == JS_INVALID_REFERENCE) {
+  if (metadata == JS_INVALID_REFERENCE) {
+    return Local<Value>();
+  }
```
Comment (kunalspathak) [inline deps/chakrashim/src/v8trycatch.cc]: Why not merge both if-conditions?
```C++
  if (metadata == JS_INVALID_REFERENCE) {
    GetAndClearException();
    return Local<Value>();
  }
```

Comment (MSLaguana) [reply deps/chakrashim/src/v8trycatch.cc]: The `GetAndClearException()` function sets the `metadata` property, unless it fails. The first check is for whether we need to initialize the `TryCatch`, the second is to see whether it succeeded.

Comment (kunalspathak) [reply deps/chakrashim/src/v8trycatch.cc]: never mind, i see that `metadata` is updated inside `GetAndClearException()`. I misread it to be calling `JsGetAndClearException()`.

---
In reply to: [118545047](https://github.com/nodejs/node-chakracore/pull/263#discussion_r118545047) [](ancestors = 118545047)

Comment (kfarnung) [reply deps/chakrashim/src/v8trycatch.cc]: Admittedly I find this pattern to be unintuitive, seems like maybe the signature/name of the method could be updated to make it more clear?

File: deps/chakrashim/src/v8message.cc (hunk 2 above)
Comment (kunalspathak) [reply deps/chakrashim/src/v8message.cc]: yeah, i agree. Please follow the indentation we use in other files.

---
In reply to: [118531320](https://github.com/nodejs/node-chakracore/pull/263#discussion_r118531320) [](ancestors = 118531320)

**Comment by MSLaguana** [comment]
CI: https://ci.nodejs.org/job/chakracore-test-pull-request/53/

File: deps/chakrashim/include/v8.h [hunk 8]
```diff
@@ -2736,8 +2737,10 @@ class V8_EXPORT TryCatch {
   void SetNonUser() { user = false; }
   void GetAndClearException();
   void CheckReportExternalException();
+  JsValueRef EnsureException() const;
```
Comment (MSLaguana) [inline deps/chakrashim/include/v8.h:2740]: What are thoughts on this? The `const` is a bold faced lie, but some of the methods that call it are `const` and the `const_cast` was there before.

**Review by kunalspathak — APPROVED** [review]
LGTM, if CI is green.

**Pull request closed without merging**